include/openthread/coap.h

```diff
@@ -468,6 +470,36 @@ otError otCoapMessageAppendObserveOption(otMessage *aMessage, uint32_t aObserve)
  */
 otError otCoapMessageAppendUriPathOptions(otMessage *aMessage, const char *aUriPath);
 
+/**
+ * This function appends a Block2 option
+ *
+ * @param[inout]  aMessage  A pointer to the CoAP message.
+ * @param[in]     aNum      Current block number.
+ * @param[in]     aMore     Boolean to indicate more blocks are to be sent.
+ * @param[in]     aSize     Maximum block size.
+ *
+ * @retval OT_ERROR_NONE          Successfully appended the option.
+ * @retval OT_ERROR_INVALID_ARGS  The option type is not equal or greater than the last option type.
+ * @retval OT_ERROR_NO_BUFS       The option length exceeds the buffer size.
+ *
+ */
+otError otCoapMessageAppendBlock2Option(otMessage *aMessage, uint32_t aNum, bool aMore, uint16_t aSize);
```

Comment:
The implementation expects `aSize` to be certain values 16, 32, 64, ..., 1024. All other values would cause a `OT_ERROR_INVALID_ARGS` error.

How about instead of using a `uint16_t aSize`, we define an `enum otCoapBlockSize` type:
```
typedef enum otCoapBlockSize
{
    OT_COAP_BLOCK_SIZE_16 = 0,
    OT_COAP_BLOCK_SIZE_32 = 1,
    ...
} otCoapBlockSize;
```
I feel this way it is more clear how the API is intended to be used.  Also the enum values  already map to encoded value which would also simplify the encoding logic in `AppendBlockOption()`

Thoughts?

Reply:
Yeah, I like that solution much better than passing in an arbitrary int. I'll make the change!

Reply:
I just thought of a few downsides to using an enum. Most applications will still need to have an additional integer value of the block size around in order to perform data segmentation and block number calculations, which is kind of annoying. Also, with a C compiler, a user won't get a warning/error when supplying an invalid integer instead of an enum member.

I still think your proposed solution is more elegant, however.

Reply:
The other way would be for the application to specify the size exponent directly, and instead provide a means (macro, inline function) to convert between that and and nearest size… something like:

```
#define OT_COAP_MIN_SZX (0)
#define OT_COAP_MAX_SZX (7)

uint16_t otCoapBlockSizeFromExponent(uint8_t aSzx)
{
    return 1 << (aSzx + 4);
}

uint8_t otCoapNearestSizeExponent(uint16_t aBlockSize)
{
    uint8_t szx;
    for (szx = OT_COAP_MIN_SZX; szx < OT_COAP_MAX_SZX; szx++)
    {
        if (otCoapBlockSizeFromExponent(szx) >= aBlockSize)
        {
            break;
        }
    }

    return szx;
}
```

That will probably need some tweaking, and there may be optimisations that can be made (I took the naïve iterative approach to illustrate), but that will handle arbitrary sizes like 192 bytes, by rounding up to the nearest block size (256 in that case).

Then application writers can bounce either way.  The C compiler may even be smart enough if given a constant to `otCoapNearestSizeExponent`, that the result never changes and so replace the function call with a constant.  The return value of `otCoapNearestSizeExponent` could possibly be cast to that `enum` as well.

Reply:
Yeah, I was thinking at the very least that providing a helper function to convert to/from an encoded size exponent would be helpful. I think a user providing a non-existent size should result in an error. I don't think we should increase/decrease size to a valid number. The interface can and should be well defined and limited - a user should supply a valid size.

Reply:
Fair enough, thankfully there aren't lots of options (8 possible values, the powers of two between 16 and 2048… we might omit 2048 since the MTU won't support it).

The complicating matter is you then can't do something like…
```
otCoapMessageAppendBlock2Option(mymessage, mynum, mynum < mylast, otCoapGimmieMySzx(1024));
```
…which would work as expected otherwise, unless we added a `OT_COAP_BLOCK_SIZE_INVALID` entry to the size `enum` (which has its own problems).

I'd say the case where the application chooses the `SZX` directly and needs to know the block size is more probable though, and for that, `otCoapBlockSizeFromExponent` alone would suffice.

So one approach is we provide a convenience function to go one way, and let the application developer beat a path back if they want to do it the other way.